### PR TITLE
Add support for syntax introduced in JDK10 and JDK11

### LINKED ISSUE #74: Clover fails to parse try-with-resources when closeable resource has no type specified
Example from OpenJDK 9, FileDescriptor:

```
@SuppressWarnings("try")
    synchronized void closeAll(Closeable releaser) throws IOException {
        if (!closed) {
            closed = true;
            IOException ioe = null;
            try (releaser) {   // <<<<<<<<<<<<
                if (otherParents != null) {
                    for (Closeable referent : otherParents) {
                        try {
                            referent.close();
                        } catch(IOException x) {
                            if (ioe == null) {
                                ioe = x;
                            } else {
                                ioe.addSuppressed(x);
                            }
                        }
                    }
                }
            } catch(IOException ex) {
                /*
                 * If releaser close() throws IOException
                 * add other exceptions as suppressed.
                 */
                if (ioe != null)
                    ex.addSuppressed(ioe);
                ioe = ex;
            } finally {
                if (ioe != null)
                    throw ioe;
            }
        }
    }```

it seems it fails on "try (variable) {" - java.g grammar defines

"try" LPAREN declaration ...

declaration: fieldModifiers typeSpec variableDefinitions

Workaround:

Exclude entire file from instrumentation

Expected fix:

Allow variable name (names?), without types.

### LINKED ISSUE #136: MismatchedTokenException for try-with-resources
I ran into an exception today with openclover:clover-maven-plugin:4.4.1 running on an Azure Pipelines CI job, and could strip the class in question down to this:
```java
import java.io.*;

public class Test {
    private final OutputStream out = new ByteArrayOutputStream();

    private void test() {
        try (out) {
        } catch (Throwable t) {
        }
    }
}
```
Above class will lead to the following error:
```
com.atlassian.clover.api.CloverException: /home/vsts/work/1/s/src/main/java/Test.java:7:9:unexpected token: try
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:159)
    at com.atlassian.clover.CloverInstr.execute (CloverInstr.java:110)
    at com.atlassian.clover.CloverInstr.mainImpl (CloverInstr.java:86)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: clover.antlr.NoViableAltException: unexpected token: try
    at com.atlassian.clover.instr.java.JavaRecognizer.statement (JavaRecognizer.java:3705)
    at com.atlassian.clover.instr.java.JavaRecognizer.outerCompoundStmt (JavaRecognizer.java:3197)
    at com.atlassian.clover.instr.java.JavaRecognizer.field (JavaRecognizer.java:2766)
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:211)
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:124)
    at com.atlassian.clover.CloverInstr.execute (CloverInstr.java:110)
    at com.atlassian.clover.CloverInstr.mainImpl (CloverInstr.java:86)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrumentSources (AbstractInstrumenter.java:205)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrument (AbstractInstrumenter.java:80)
    at com.atlassian.maven.plugin.clover.CloverInstrumentInternalMojo.execute (CloverInstrumentInternalMojo.java:279)
    at com.atlassian.maven.plugin.clover.CloverSetupMojo.execute (CloverSetupMojo.java:31)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
...
[ERROR] Failed to execute goal org.openclover:clover-maven-plugin:4.4.1:setup (default-cli) on project kafka-commons: Clover has failed to instrument the source files in the [/home/vsts/work/1/s/target/clover/src-instrumented] directory -> [Help 1]
```

### PR DESCRIPTION
**Pull request by satob**
Closes #74, closes #136
- Add testcase for syntax introduced in Java 10 (var variable)
- Add support for syntax introduced in Java 11 (try-with-resource with predefined variables, var in parameter of lambda)

### CODE REVIEW

**Comment by satob** [comment]
This pull request also closes #74 .

File: clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java [hunk 1]
```diff
@@ -20,6 +23,7 @@ public class JavaEnvUtils {
             javaVersion = JAVA_8;
             Class.forName("java.lang.StackWalker");
             javaVersion = JAVA_9;
+            javaVersion = SourceVersion.latestSupported().toString().substring("RELEASE_".length());
```
Comment (marek-parfianowicz) [inline clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java:26]: I would like to have code detection based on the Java API. Therefore, please find some classes, preferably from java.core, which have been introduced in Java 10 and Java 11, respectively.

File: clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax18CompilationTest.groovy [hunk 2]
```diff
@@ -470,8 +470,11 @@ class JavaSyntax18CompilationTest extends JavaSyntaxCompilationTestBase {
     }
 
     void testPackageInfo() {
-        if (JavaEnvUtils.isAtLeastJavaVersion(JavaEnvUtils.JAVA_1_8)) {
+        // @XmlAccessorType annotation is deprecated in JDK 9 and not available in JDK 11 or later
+        if (JavaEnvUtils.getJavaVersion().equals("8")) {
             instrumentAndCompileSourceFile(srcDir, mGenSrcDir, "pck/package-info.java", JavaEnvUtils.JAVA_1_8)
+        } else if (JavaEnvUtils.isAtLeastJavaVersion(JavaEnvUtils.JAVA_9)) {
+            instrumentAndCompileSourceFile(srcDir, mGenSrcDir, "pck2/package-info.java", JavaEnvUtils.JAVA_1_8)
```
Comment (marek-parfianowicz) [inline clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax18CompilationTest.groovy:477]:
```suggestion
            instrumentAndCompileSourceFile(srcDir, mGenSrcDir, "pck2/package-info.java", JavaEnvUtils.JAVA_1_9)
```

File: clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java (hunk 1 above)
Comment (satob) [reply clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java]: I have changed the way to get the Java runtime version to use classes/methods introduced in each version.

**Pull request merged**